File: server/service/src/sync/test/test_data/reason.rs
use crate::sync::test::TestSyncIncomingRecord;
use repository::{ReasonOptionRow, ReasonOptionType};

const INVENTORY_ADJUSTMENT_REASON_1: (&str, &str) = (
    "positive_adjustment",
    r#"{
    "ID": "positive_adjustment",
    "type": "positiveInventoryAdjustment",
    "isActive": true,
    "title": "Found"
    }"#,
);
const RETURN_REASON_1: (&str, &str) = (
    "return_reason",
    r#"{
    "ID": "return_reason",
    "type": "returnReason",
    "isActive": true,
    "title": "Damaged"
    }"#,
);

pub(crate) fn test_pull_upsert_records() -> Vec<TestSyncIncomingRecord> {
    vec![
        TestSyncIncomingRecord::new_pull_upsert(
            "options",
            INVENTORY_ADJUSTMENT_REASON_1,
            ReasonOptionRow {
                id: INVENTORY_ADJUSTMENT_REASON_1.0.to_string(),
                r#type: ReasonOptionType::PositiveInventoryAdjustment,
                is_active: true,
                reason: "Found".to_string(),
            },
        ),
        TestSyncIncomingRecord::new_pull_upsert(
            "options",
            RETURN_REASON_1,
            ReasonOptionRow {
                id: RETURN_REASON_1.0.to_string(),
                r#type: ReasonOptionType::ReturnReason,
                is_active: true,
                reason: "Damaged".to_string(),
            },
        ),
    ]
}

File: server-go/internal/sync/translations/reason.go
package translations

import (
	"encoding/json"
	"fmt"

	"github.com/msupply-foundation/open-msupply/server-go/internal/db"
	"github.com/msupply-foundation/open-msupply/server-go/internal/repository"
	"github.com/msupply-foundation/open-msupply/server-go/internal/sync/synctypes"
)

// legacyOptionsRow mirrors translations::reason::LegacyOptionsRow. The legacy sync table is
// "options"; the "type" field uses camelCase enum values.
type legacyOptionsRow struct {
	ID       string `json:"ID"`
	Type     string `json:"type"`
	IsActive bool   `json:"isActive"`
	Reason   string `json:"title"`
}

// legacyReasonType maps the legacy camelCase options type to the repository ReasonOptionType.
var legacyReasonType = map[string]repository.ReasonOptionType{
	"positiveInventoryAdjustment": repository.ReasonPositiveInventoryAdjustment,
	"negativeInventoryAdjustment": repository.ReasonNegativeInventoryAdjustment,
	"openVialWastage":             repository.ReasonOpenVialWastage,
	"returnReason":                repository.ReasonReturnReason,
	"requisitionLineVariance":     repository.ReasonRequisitionLineVariance,
	"closedVialWastage":           repository.ReasonClosedVialWastage,
}

// Reason is a trivial, dependency-free, untracked translator (legacy table "options").
type Reason struct{ synctypes.BaseTranslation }

func (Reason) TableNames() []string { return []string{"options"} }

func (Reason) TryTranslateFromUpsert(_ synctypes.Exec, _ db.Dialect, row *synctypes.SyncBufferRow) (synctypes.PullTranslateResult, error) {
	var data legacyOptionsRow
	if err := json.Unmarshal([]byte(row.Data), &data); err != nil {
		return synctypes.NotMatched(), err
	}
	rt, ok := legacyReasonType[data.Type]
	if !ok {
		return synctypes.NotMatched(), fmt.Errorf("unknown options type %q for reason %s", data.Type, data.ID)
	}
	return synctypes.PullUpsert(repository.ReasonOptionRow{
		ID:       data.ID,
		Type:     rt,
		IsActive: data.IsActive,
		Reason:   data.Reason,
	}), nil
}

func (Reason) TryTranslateFromDelete(_ synctypes.Exec, _ db.Dialect, row *synctypes.SyncBufferRow) (synctypes.PullTranslateResult, error) {
	return synctypes.PullDelete(repository.ReasonOptionRowDelete{ID: row.RecordID}), nil
}
